pkg/envoy/lds/builder.go

```diff
@@ -0,0 +1,599 @@
+	fb.statsPrefix = statsPrefix
+	return fb
+}
+
+// WithRBAC sets the RBAC properties used to build the filter
+func (fb *filterBuilder) WithRBAC(t []trafficpolicy.TrafficTargetWithRoutes, trustDomain string) *filterBuilder {
+	fb.withRBAC = true
+	fb.trafficTargets = t
+	fb.trustDomain = trustDomain
+	return fb
+}
+
+func (fb *filterBuilder) TCPLocalRateLimit(rl *policyv1alpha1.TCPLocalRateLimitSpec) *filterBuilder {
+	fb.tcpLocalRateLimit = rl
+	return fb
+}
+
+func (fb *filterBuilder) TCPGlobalRateLimit(rl *policyv1alpha1.TCPGlobalRateLimitSpec) *filterBuilder {
+	fb.tcpGlobalRateLimit = rl
+	return fb
+}
+
+func (fb *filterBuilder) HTTPConnManager() *httpConnManagerBuilder {
+	if fb.hcmBuilder == nil {
+		fb.hcmBuilder = HTTPConnManagerBuilder()
+	}
+
+	return fb.hcmBuilder
+}
+
+func (fb *filterBuilder) TCPProxy() *tcpProxyBuilder {
+	if fb.tcpProxyBuilder == nil {
+		fb.tcpProxyBuilder = TCPProxyBuilder()
+	}
+	return fb.tcpProxyBuilder
+}
+
+func (fb *filterBuilder) Build() ([]*xds_listener.Filter, error) {
+	var filters []*xds_listener.Filter
+
+	// RBAC filter should be the very first filter in the filter chain
+	if fb.withRBAC {
+		rbacFilter, err := buildRBACFilter(fb.trafficTargets, fb.trustDomain)
+		if err != nil {
+			return nil, err
+		}
+		filters = append(filters, rbacFilter)
+	}
+
+	// Rate limit filters
+	if fb.tcpLocalRateLimit != nil {
+		rateLimitFilter, err := buildTCPLocalRateLimitFilter(fb.tcpLocalRateLimit, fb.statsPrefix)
+		if err != nil {
+			return nil, err
+		}
+		filters = append(filters, rateLimitFilter)
+	}
+	if fb.tcpGlobalRateLimit != nil {
+		rateLimitFilter, err := buildTCPGlobalRateLimitFilter(fb.tcpGlobalRateLimit, fb.statsPrefix)
+		if err != nil {
+			return nil, err
+		}
+		filters = append(filters, rateLimitFilter)
+	}
+
+	// HTTP connection manager filter
+	if fb.hcmBuilder != nil {
+		hcmFilter, err := fb.hcmBuilder.Build()
+		if err != nil {
+			return nil, err
+		}
+		filters = append(filters, hcmFilter)
+	}
+
+	if fb.tcpProxyBuilder != nil {
+		tcpProxyFilter, err := fb.tcpProxyBuilder.Build()
+		if err != nil {
+			return nil, err
+		}
+		filters = append(filters, tcpProxyFilter)
+	}
+
+	return filters, nil
+}
+
+// HTTPConnManagerBuilder returns the HTTP Connection Manager builder instance
+func HTTPConnManagerBuilder() *httpConnManagerBuilder { //nolint: revive // unexported-return
+	return &httpConnManagerBuilder{}
+}
+
+// StatsPrefix sets the stats prefix on the builder
+func (hb *httpConnManagerBuilder) StatsPrefix(statsPrefix string) *httpConnManagerBuilder {
+	hb.statsPrefix = statsPrefix
+	return hb
+}
+
+// RouteConfigName sets the route config name on the builder
+func (hb *httpConnManagerBuilder) RouteConfigName(name string) *httpConnManagerBuilder {
+	hb.routeConfigName = name
+	return hb
+}
+
+// DefaultFilters sets the default HTTP filters on the builder
+func (hb *httpConnManagerBuilder) DefaultFilters() *httpConnManagerBuilder {
```

Comment:
Nope, the order of filters matters and this should be called in the current order.

Reply:
but since this is always called first, doesn't that leave potential for ommitting it and therefore incorrect behavior? Why not bring it into the build method and remove that possibility? Looks like it's an implementation detail to me and shouldn't be exported

Reply:
Updated to `defaultFilters()` to handle the ordering logic within `build()`.